File: crates/runlet-core/src/js/text.js
(function () {
  // The first-class `text` value-util (beside `$`/`money`/`Decimal`/`datetime`). An immutable
  // string value with a snake_case, Python-flavored author surface: the method NAMES are Pythonic
  // renames of native JS string operations — the SEMANTICS are JavaScript's (UTF-16 code units for
  // counting/width; Unicode-default, locale-independent casing). It is pure — no clock, no
  // randomness, no ambient state — so it is injected identically under every profile and there is
  // nothing for the determinism sanitizer to remove. No `__sys` bridge and no Rust math: everything
  // here composes from `String.prototype`.

  // Caller-controlled width/repeat is capped before allocation so `text("x").rjust(1e9)` cannot OOM
  // the isolate — the same fail-closed spirit as the engine's `max_*_bytes` limits.
  var MAX_OUTPUT = 1000000; // 1,000,000 UTF-16 code units

  function Text(s) {
    this.value = s; // the underlying plain string
  }

  function isText(x) { return x instanceof Text; }

  // Coerce any operand (a text value or a raw value) to a plain string.
  function strOf(x) { return isText(x) ? x.value : String(x); }

  function wrap(s) { return new Text(s); }

  // Guard a to-be-produced length against the output cap.
  function checkWidth(n) {
    if (n > MAX_OUTPUT) {
      throw new Error("text: requested width " + n + " exceeds max " + MAX_OUTPUT);
    }
  }

  // ---- case (renamed, Unicode-default) ----------------------------------
  Text.prototype.lower = function () { return wrap(this.value.toLowerCase()); };
  Text.prototype.upper = function () { return wrap(this.value.toUpperCase()); };
  Text.prototype.capitalize = function () {
    var s = this.value;
    return wrap(s ? s.charAt(0).toUpperCase() + s.slice(1).toLowerCase() : s);
  };
  // title-case: capitalize each whitespace-separated word (native has no title-case).
  Text.prototype.title = function () {
    return wrap(this.value.replace(/\S+/g, function (w) {
      return w.charAt(0).toUpperCase() + w.slice(1).toLowerCase();
    }));
  };
  Text.prototype.swap_case = function () {
    return wrap(this.value.replace(/[a-zA-Z]/g, function (c) {
      return c === c.toLowerCase() ? c.toUpperCase() : c.toLowerCase();
    }));
  };

  // ---- strip (optional char set, else whitespace) -----------------------
  // Build a character class from an explicit set of strip characters.
  function stripClass(chars) {
    var escaped = chars.replace(/[.*+?^${}()|[\]\\-]/g, "\\$&");
    return "[" + escaped + "]";
  }
  Text.prototype.strip = function (chars) {
    if (chars === undefined) return wrap(this.value.trim());
    var re = new RegExp("^" + stripClass(chars) + "+|" + stripClass(chars) + "+$", "g");
    return wrap(this.value.replace(re, ""));
  };
  Text.prototype.lstrip = function (chars) {
    if (chars === undefined) return wrap(this.value.replace(/^\s+/, ""));
    return wrap(this.value.replace(new RegExp("^" + stripClass(chars) + "+"), ""));
  };
  Text.prototype.rstrip = function (chars) {
    if (chars === undefined) return wrap(this.value.replace(/\s+$/, ""));
    return wrap(this.value.replace(new RegExp(stripClass(chars) + "+$"), ""));
  };

  // ---- prefix / suffix --------------------------------------------------
  Text.prototype.starts_with = function (p) { return this.value.indexOf(strOf(p)) === 0; };
  Text.prototype.ends_with = function (s) {
    var v = this.value, suf = strOf(s);
    return v.indexOf(suf, v.length - suf.length) !== -1 && v.length >= suf.length;
  };
  Text.prototype.removeprefix = function (p) {
    var pre = strOf(p);
    return wrap(this.value.indexOf(pre) === 0 ? this.value.slice(pre.length) : this.value);
  };
  Text.prototype.removesuffix = function (s) {
    var suf = strOf(s), v = this.value;
    var has = suf.length > 0 && v.indexOf(suf, v.length - suf.length) === v.length - suf.length;
    return wrap(has ? v.slice(0, v.length - suf.length) : v);
  };

  // ---- replace (replaces ALL, matching Python's str.replace) ------------
  Text.prototype.replace = function (old, neu) {
    var from = strOf(old);
    if (from === "") return wrap(this.value);
    return wrap(this.value.split(from).join(strOf(neu)));
  };

  // ---- count (non-overlapping; empty needle → len+1, like Python) -------
  Text.prototype.count = function (sub) {
    var needle = strOf(sub);
    if (needle === "") return this.value.length + 1;
    return this.value.split(needle).length - 1;
  };

  // ---- split / splitlines (return plain strings) ------------------------
  // Left-to-right split with optional Python-style maxsplit (remainder kept in the last piece).
  Text.prototype.split = function (sep, maxsplit) {
    var s = this.value, by = strOf(sep);
    if (maxsplit === undefined || maxsplit < 0) return s.split(by);
    var out = [], start = 0, n = 0;
    while (n < maxsplit) {
      var idx = s.indexOf(by, start);
      if (idx === -1) break;
      out.push(s.slice(start, idx));
      start = idx + by.length;
      n++;
    }
    out.push(s.slice(start));
    return out;
  };
  // Right-to-left split; without maxsplit it equals split.
  Text.prototype.rsplit = function (sep, maxsplit) {
    var s = this.value, by = strOf(sep);
    if (maxsplit === undefined || maxsplit < 0) return s.split(by);
    var out = [], end = s.length, n = 0;
    while (n < maxsplit) {
      var idx = s.lastIndexOf(by, end - by.length);
      if (idx === -1 || by.length === 0) break;
      out.unshift(s.slice(idx + by.length, end));
      end = idx;
      n++;
    }
    out.unshift(s.slice(0, end));
    return out;
  };
  Text.prototype.splitlines = function () {
    if (this.value === "") return [];
    return this.value.split(/\r\n|\r|\n/);
  };

  // ---- padding / alignment (bounded) ------------------------------------
  // Python-style zero-fill honoring a leading sign: "-42".zfill(5) → "-0042".
  Text.prototype.zfill = function (width) {
    var w = Number(width);
    checkWidth(w);
    var s = this.value;
    if (s.length >= w) return wrap(s);
    var sign = (s.charAt(0) === "+" || s.charAt(0) === "-") ? s.charAt(0) : "";
    var body = sign ? s.slice(1) : s;
    var pad = w - s.length;
    return wrap(sign + new Array(pad + 1).join("0") + body);
  };
  Text.prototype.ljust = function (width, fill) {
    var w = Number(width);
    checkWidth(w);
    return wrap(this.value.padEnd(w, fill === undefined ? " " : String(fill)));
  };
  Text.prototype.rjust = function (width, fill) {
    var w = Number(width);
    checkWidth(w);
    return wrap(this.value.padStart(w, fill === undefined ? " " : String(fill)));
  };
  Text.prototype.center = function (width, fill) {
    var w = Number(width);
    checkWidth(w);
    var s = this.value;
    if (s.length >= w) return wrap(s);
    var f = fill === undefined ? " " : String(fill);
    var total = w - s.length;
    var left = Math.floor(total / 2);
    var right = total - left;
    return wrap(new Array(left + 1).join(f) + s + new Array(right + 1).join(f));
  };

  // ---- character-class predicates (non-empty; Unicode-aware) ------------
  Text.prototype.is_digit = function () { return /^\p{Nd}+$/u.test(this.value); };
  Text.prototype.is_alpha = function () { return /^\p{L}+$/u.test(this.value); };
  Text.prototype.is_alnum = function () { return /^[\p{L}\p{Nd}]+$/u.test(this.value); };
  Text.prototype.is_space = function () { return this.value.length > 0 && /^\s+$/.test(this.value); };

  // ---- ERP shaping verbs ------------------------------------------------
  // NFD-fold diacritics → lowercase → collapse non-alphanumerics to single hyphens, no edge hyphens.
  Text.prototype.slugify = function () {
    var s = this.value.normalize("NFD").replace(/[̀-ͯ]/g, "");
    s = s.toLowerCase().replace(/[^a-z0-9]+/g, "-").replace(/^-+|-+$/g, "");
    return wrap(s);
  };
  // Lossy DISPLAY masking (not encoding): keep a tail, mask the rest. Default keep 4, char "*".
  Text.prototype.mask = function (opts) {
    var o = opts || {};
    var keep = o.keep === undefined ? 4 : Number(o.keep);
    var ch = o.char === undefined ? "*" : String(o.char).charAt(0) || "*";
    var s = this.value;
    if (keep <= 0) return wrap(new Array(s.length + 1).join(ch));
    if (keep >= s.length) return wrap(s);
    var masked = new Array(s.length - keep + 1).join(ch);
    return wrap(masked + s.slice(s.length - keep));
  };
  Text.prototype.redact = Text.prototype.mask;
  // Trim then collapse internal whitespace runs to single spaces.
  Text.prototype.collapse = function () {
    return wrap(this.value.replace(/\s+/g, " ").trim());
  };
  // Shorten to at most `limit` code units; append the ellipsis marker (counted toward the limit)
  // when truncation happens.
  Text.prototype.truncate = function (limit, opts) {
    var lim = Number(limit);
    var mark = (opts && opts.ellipsis !== undefined) ? String(opts.ellipsis) : "…";
    var s = this.value;
    if (s.length <= lim) return wrap(s);
    if (lim <= 0) return wrap("");
    if (mark.length >= lim) return wrap(s.slice(0, lim));
    return wrap(s.slice(0, lim - mark.length) + mark);
  };

  // ---- length + interop -------------------------------------------------
  Text.prototype.len = function () { return this.value.length; };
  Text.prototype.to_string = function () { return this.value; };
  Text.prototype.toString = function () { return this.value; };
  Text.prototype.valueOf = function () { return this.value; };
  // json()/JSON.stringify serialize the plain string, never a wrapper object.
  Text.prototype.toJSON = function () { return this.value; };

  // Internal interop hooks the `list` verbs read (not author-facing, absent from base.d.ts):
  // __order_key → the string (lexical ordering); __id_key → the string (grouping/dedup/equality).
  Text.prototype.__order_key = function () { return this.value; };
  Text.prototype.__id_key = function () { return this.value; };

  // ---- factory ----------------------------------------------------------
  // text(input) wraps a value as a string; an existing text value is returned as-is.
  function make(input) {
    if (isText(input)) return input;
    return new Text(String(input === undefined || input === null ? "" : input));
  }
  globalThis.text = make;
})();

File: crates/runlet-core/src/text.rs
//! The first-class `text` value-util for the `QuickJS` sandbox.
//!
//! `text` is the always-on string value-util at `$std.text` (beside `$`/`$std.decimal`/`$std.datetime`): a callable
//! factory (`text(input)`) that wraps a value as an immutable string with chainable, `snake_case`
//! methods. The method names are Python-flavored renames of native JS string operations; the
//! semantics are JavaScript's (UTF-16 code units, Unicode-default casing). A small set of
//! ERP-common shaping verbs (`slugify`/`mask`/`collapse`/`truncate`/padding) composes from the
//! same primitives.
//!
//! Pure JS (`js/text.js`) — no `__sys` bridge and no Rust math: unlike `datetime`, every operation
//! is expressible over `String.prototype`, so this injector only evals the wrapper. It is injected
//! under **both** profiles: `text` touches no clock, no randomness, and no ambient authority, so
//! the deterministic sanitizer (`js/determinism.js`) removes nothing from it.

use std::error::Error;

use rquickjs::{Ctx, Value as JsValue};

/// JS wrapper — loaded from `src/js/text.js` at compile time. Depends on no other injected global.
const TEXT_WRAPPER: &str = include_str!("js/text.js");

/// Injects `$std.text`. Order-independent among the pure value-utils (no bridge dependency).
///
/// # Errors
///
/// Returns an error if the JS eval fails.
pub fn inject_text(qctx: &Ctx<'_>) -> Result<(), Box<dyn Error + Send + Sync>> {
    let wrapper: JsValue<'_> = qctx.eval(TEXT_WRAPPER)?;
    drop(wrapper);
    Ok(())
}

#[cfg(test)]
mod tests {
    //! Behavioral tests for the `text` value-util, driven end-to-end through the QuickJS engine
    //! (the surface is pure JS, so there is no Rust `dispatch` to unit-test directly). Each test
    //! injects the wrapper and evals an assertion expression.

    use rquickjs::{Context, Runtime};

    use super::inject_text;

    /// The deterministic-profile sanitizer, evaled to prove `text` survives it untouched.
    const DETERMINISM: &str = include_str!("js/determinism.js");

    /// Bootstraps `$std` (the wrapper now populates it, not `globalThis`), injects `text`, then
    /// mirrors it back to a bare `text` global so these behavioral expressions read naturally. In
    /// production the engine does the bootstrap + projection; here the harness stands in for it.
    fn inject(qctx: &rquickjs::Ctx<'_>) {
        qctx.eval::<(), _>("globalThis.$std = {};")
            .expect("bootstrap std");
        inject_text(qctx).expect("inject text");
        qctx.eval::<(), _>("globalThis.text = $std.text;")
            .expect("project text");
    }

    /// Inject `text` and eval a JS expression that yields a string.
    fn run(expr: &str) -> String {
        let rt = Runtime::new().expect("runtime");
        let ctx = Context::full(&rt).expect("context");
        ctx.with(|qctx| {
            inject(&qctx);
            qctx.eval::<String, _>(expr).expect("eval")
        })
    }

    #[test]
    fn unwrap_forms_yield_plain_strings() {
        assert_eq!(run(r#"text("Ac-Me").lower().value"#), "ac-me");
        assert_eq!(run(r#"String(text("Ac-Me").lower())"#), "ac-me");
        assert_eq!(run(r#"JSON.stringify(text("Ac-Me").lower())"#), "\"ac-me\"");
    }

    #[test]
    fn transforms_do_not_mutate_the_receiver() {
        // strip returns a new value; the original is unchanged.
        assert_eq!(
            run(
                r#"(function(){var t=text("  Hi  ");var s=t.strip().value;return s+"|"+t.value;})()"#
            ),
            "Hi|  Hi  "
        );
    }

    #[test]
    fn case_and_strip_renames() {
        assert_eq!(run(r#"text("  Héllo  ").strip().upper().value"#), "HÉLLO");
        assert_eq!(run(r#"text("HeLLo").swap_case().value"#), "hEllO");
        assert_eq!(run(r#"text("mixED cAse").title().value"#), "Mixed Case");
        assert_eq!(run(r#"text("hello").capitalize().value"#), "Hello");
        // strip with an explicit character set.
        assert_eq!(run(r#"text("xxabcxx").strip("x").value"#), "abc");
    }

    #[test]
    fn prefix_suffix_and_predicates() {
        assert_eq!(
            run(r#"String(text("SKU-0042").starts_with("SKU-"))"#),
            "true"
        );
        assert_eq!(
            run(r#"text("SKU-0042").removeprefix("SKU-").value"#),
            "0042"
        );
        assert_eq!(
            run(r#"text("file.txt").removesuffix(".txt").value"#),
            "file"
        );
        assert_eq!(run(r#"String(text("0042").is_digit())"#), "true");
        assert_eq!(run(r#"String(text("Café").is_alpha())"#), "true");
        assert_eq!(run(r#"String(text("A1").is_alnum())"#), "true");
        assert_eq!(run(r#"String(text("").is_digit())"#), "false");
        assert_eq!(run(r#"String(text("  ").is_space())"#), "true");
    }

    #[test]
    fn replace_and_count() {
        // replace hits ALL occurrences (Python semantics), unlike native String.replace.
        assert_eq!(run(r#"text("a.b.c").replace(".", "-").value"#), "a-b-c");
        assert_eq!(run(r#"String(text("a.b.c").count("."))"#), "2");
    }

    #[test]
    fn splitting_returns_plain_strings() {
        assert_eq!(
            run(r#"JSON.stringify(text("a,b,c").split(","))"#),
            r#"["a","b","c"]"#
        );
        assert_eq!(
            run(r#"JSON.stringify(text("a\nb").splitlines())"#),
            r#"["a","b"]"#
        );
        // maxsplit keeps the remainder in the last piece (Python str.split).
        assert_eq!(
            run(r#"JSON.stringify(text("a,b,c").split(",",1))"#),
            r#"["a","b,c"]"#
        );
        assert_eq!(
            run(r#"JSON.stringify(text("a,b,c").rsplit(",",1))"#),
            r#"["a,b","c"]"#
        );
    }

    #[test]
    fn padding_and_alignment() {
        assert_eq!(run(r#"text("42").zfill(6).value"#), "000042");
        assert_eq!(run(r#"text("-42").zfill(6).value"#), "-00042");
        assert_eq!(run(r#"text("x").rjust(5).value"#), "    x");
        assert_eq!(run(r#"text("x").ljust(3, ".").value"#), "x..");
        assert_eq!(run(r#"text("x").center(5, "-").value"#), "--x--");
    }

    #[test]
    fn oversize_width_is_refused() {
        // width beyond the output cap throws rather than allocating unboundedly.
        assert_eq!(
            run(
                r#"(function(){try{text("x").rjust(2e9);return "no-throw";}catch(e){return "threw";}})()"#
            ),
            "threw"
        );
    }

    #[test]
    fn erp_verbs() {
        assert_eq!(
            run(r#"text("  Café Ör 01! ").slugify().value"#),
            "cafe-or-01"
        );
        assert_eq!(
            run(r#"text("4111111111111234").mask().value"#),
            "************1234"
        );
        assert_eq!(
            run(r##"text("4111111111111234").mask({keep:4,char:"#"}).value"##),
            "############1234"
        );
        assert_eq!(run(r#"text("a   b\t c").collapse().value"#), "a b c");
        assert_eq!(run(r#"text("hello world").truncate(8).value"#), "hello w…");
        // no truncation when it already fits.
        assert_eq!(run(r#"text("short").truncate(10).value"#), "short");
    }

    #[test]
    fn present_and_identical_under_the_deterministic_sanitizer() {
        // Eval the sanitizer after injecting text, then confirm the whole surface still works —
        // text touches no ambient authority, so the deterministic profile removes nothing from it.
        let rt = Runtime::new().expect("runtime");
        let ctx = Context::full(&rt).expect("context");
        let out = ctx.with(|qctx| {
            inject(&qctx);
            qctx.eval::<(), _>(DETERMINISM).expect("eval determinism");
            qctx.eval::<String, _>(
                r#"text("  Héllo  ").strip().slugify().mask({keep:2}).value + "|" + String(typeof text)"#,
            )
            .expect("eval")
        });
        assert_eq!(out, "***lo|function");
    }
}
